Orange/evaluation/testing.py

```diff
@@ -289,58 +347,30 @@ def __init__(self, data, learners, k=10, random_state=0, store_data=False,
         self.k = k
         self.random_state = random_state
-        Y = data.Y.copy().flatten()
-        indices = None
+
+        self.indices = None
         if data.domain.has_discrete_class:
-            indices = skl_cross_validation.StratifiedKFold(
-                Y, self.k, shuffle=True, random_state=self.random_state
+            self.indices = skl_cross_validation.StratifiedKFold(
+                data.Y, self.k, shuffle=True, random_state=self.random_state
             )
-            if any(len(train) == 0 or len(test) == 0 for train, test in indices):
+            if any(len(train) == 0 or len(test) == 0 for train, test in self.indices):
                 if warnings is not None:
                     warnings.append("Using non-stratified sampling.")
-                indices = None
-        if indices is None:
-            indices = skl_cross_validation.KFold(
-                len(Y), self.k, shuffle=True, random_state=self.random_state
+                self.indices = None
+        if self.indices is None:
+            self.indices = skl_cross_validation.KFold(
+                len(data.Y), self.k, shuffle=True, random_state=self.random_state
             )
 
-        self.folds = []
-        if self.store_models:
-            self.models = []
-        ptr = 0
-        nmethods = len(learners)
-        n_callbacks = nmethods * self.k
-        for fold_idx, (train, test) in enumerate(indices):
-            train_data, test_data = data[train], data[test]
-            if len(train_data) == 0 or len(test_data) == 0:
-                raise RuntimeError("One of the train or test folds is empty.")
-            if self.preprocessor is not None:
-                train_data = self.preprocessor(train_data)
-            fold_slice = slice(ptr, ptr + len(test))
-            self.folds.append(fold_slice)
-            self.row_indices[fold_slice] = test
-            self.actual[fold_slice] = test_data.Y.flatten()
-            if self.store_models:
-                fold_models = [None] * nmethods
-                self.models.append(fold_models)
-            for i, learner in enumerate(learners):
-                model = self.train_if_succ(i, learner, train_data)
-                self.call_callback((fold_idx * nmethods + i) / n_callbacks)
-                if not model:
-                    continue
-                if self.store_models:
-                    fold_models[i] = model
-                if data.domain.has_discrete_class:
-                    values, probs = model(test_data, model.ValueProbs)
-                    self.predicted[i][fold_slice] = values
-                    self.probabilities[i][fold_slice, :] = probs
-                elif data.domain.has_continuous_class:
-                    values = model(test_data, model.Value)
-                    self.predicted[i][fold_slice] = values
-            ptr += len(test)
-        self.call_callback(1)
+        if self.preprocessor is not None:
+            data = self.preprocessor(data)
```

Comment:
Preprocessor has to be called on each training set separately, not on the entire data set. Preprocessing is considered a part of the learning and should not be able to "see" and use the test data.